File: py/testgen/fuzzjose/ecdsa_test.py
# Copyright 2017 Google Inc. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
#
# Run as:
#     sage -python ecdsa_test.py
"""Tests for testgen.ecdsa."""

import binascii
import ecdsa
import ecutil
import util


class TestVector(object):

  def __init__(self, curve, msg, qx, qy, r, s, hash_name):
    self.curve = curve
    self.msg = binascii.unhexlify(msg)
    self.qx = int(qx, 16)
    self.qy = int(qy, 16)
    self.r = int(r, 16)
    self.s = int(s, 16)
    self.hash_name = hash_name


test_vectors = [
    TestVector(
        ecutil.curveP256,
        '5905238877c77421f73e43ee3da6f2d9e2ccad5fc942dcec0cbd25482935faaf416983'
        'fe165b1a045ee2bcd2e6dca3bdf46c4310a7461f9a37960ca672d3feb5473e253605fb'
        '1ddfd28065b53cb5858a8ad28175bf9bd386a5e471ea7a65c17cc934a9d791e91491eb'
        '3754d03799790fe2d308d16146d5c9b0d0debd97d79ce8',
        '1ccbe91c075fc7f4f033bfa248db8fccd3565de94bbfb12f3c59ff46c271bf83',
        'ce4014c68811f9a21a1fdb2c0e6113e06db7ca93b7404e78dc7ccd5ca89a4ca9',
        'f3ac8061b514795b8843e3d6629527ed2afd6b1f6a555a7acabb5e6f79c8c2ac',
        '8bf77819ca05a6b2786c76262bf7371cef97b218e96f175a3ccdda2acc058903',
        'SHA-256'),
    TestVector(
        ecutil.curveP384,
        '6b45d88037392e1371d9fd1cd174e9c1838d11c3d6133dc17e65fa0c485dcca9f52d41'
        'b60161246039e42ec784d49400bffdb51459f5de654091301a09378f93464d52118b48'
        'd44b30d781eb1dbed09da11fb4c818dbd442d161aba4b9edc79f05e4b7e401651395b5'
        '3bd8b5bd3f2aaa6a00877fa9b45cadb8e648550b4c6cbe',
        'c2b47944fb5de342d03285880177ca5f7d0f2fcad7678cce4229d6e1932fcac11bfc3c'
        '3e97d942a3c56bf34123013dbf',
        '37257906a8223866eda0743c519616a76a758ae58aee81c5fd35fbf3a855b7754a36d4'
        'a0672df95d6c44a81cf7620c2d',
        '50835a9251bad008106177ef004b091a1e4235cd0da84fff54542b0ed755c1d6f25160'
        '9d14ecf18f9e1ddfe69b946e32',
        '0475f3d30c6463b646e8d3bf2455830314611cbde404be518b14464fdb195fdcc92eb2'
        '22e61f426a4a592c00a6a89721', 'SHA-384'),
    TestVector(
        ecutil.curveP521,
        '9ecd500c60e701404922e58ab20cc002651fdee7cbc9336adda33e4c1088fab1964ecb'
        '7904dc6856865d6c8e15041ccf2d5ac302e99d346ff2f686531d25521678d4fd3f76bb'
        'f2c893d246cb4d7693792fe18172108146853103a51f824acc621cb7311d2463c3361e'
        'a707254f2b052bc22cb8012873dcbb95bf1a5cc53ab89f',
        '0061387fd6b95914e885f912edfbb5fb274655027f216c4091ca83e19336740fd81aed'
        'fe047f51b42bdf68161121013e0d55b117a14e4303f926c8debb77a7fdaad1',
        '00e7d0c75c38626e895ca21526b9f9fdf84dcecb93f2b233390550d2b1463b7ee3f58d'
        'f7346435ff0434199583c97c665a97f12f706f2357da4b40288def888e59e6',
        '004de826ea704ad10bc0f7538af8a3843f284f55c8b946af9235af5af74f2b76e099e4'
        'bc72fd79d28a380f8d4b4c919ac290d248c37983ba05aea42e2dd79fdd33e8',
        '0087488c859a96fea266ea13bf6d114c429b163be97a57559086edb64aed4a18594b46'
        'fb9efc7fd25d8b2de8f09ca0587f54bd287299f47b2ff124aac566e8ee3b43',
        'SHA-512')
]


def test_sign_verify():
  curve = ecutil.curveP256
  (d, qx, qy) = ecutil.ec_generate_key(curve)
  digest = util.compute_hash('SHA-256', 'Hello')
  (r, s) = ecdsa.ecdsa_sign_hash(curve, digest, d)
  assert ecdsa.ecdsa_verify_hash(curve, digest, int(qx), int(qy), int(r),
                                 int(s))
  assert not ecdsa.ecdsa_verify_hash(curve, digest + '\x00',
                                     int(qx), int(qy), int(r), int(s))


def test_test_vectors():
  for test in test_vectors:
    assert ecdsa.ecdsa_verify_hash(test.curve,
                                   util.compute_hash(test.hash_name, test.msg),
                                   test.qx, test.qy, test.r, test.s)


if __name__ == '__main__':
  test_sign_verify()
  test_test_vectors()

File: javascript/webcryptoapi/testcases/EcdsaTest.js
/**
 * @license
 * Copyright 2017 Google Inc. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Tests for ECDSA implementations of Web Crypto API.
 */
goog.provide('wycheproof.webcryptoapi.ECDSA');
goog.require('e2e.ecc.Ecdsa');
goog.require('goog.testing.TestCase');
goog.require('goog.testing.asserts');
goog.require('goog.testing.jsunit');
goog.require('wycheproof.BigInteger');
goog.require('wycheproof.EcUtil');
goog.require('wycheproof.TestUtil');
goog.require('wycheproof.webcryptoapi.HashUtil');

var TestUtil = wycheproof.TestUtil;
var EcUtil = wycheproof.EcUtil;
var BigInteger = wycheproof.BigInteger;
var HashUtil = wycheproof.webcryptoapi.HashUtil;

// Test vector files
var ECDSA_VECTOR_FILE = '../../testvectors/ecdsa_webcrypto_test.json';

/** ECDSA wrapper */
var Ecdsa = function() {};

/**
 * Verifies the given signature using the given ECDSA public key.
 * @param {!CryptoKey} pk The ECDSA public key
 * @param {string} hashAlg The hash algorithm
 * @param {!ArrayBuffer} msg The message to be verified
 * @param {!ArrayBuffer} sig The signature to be verified
 *
 * @return {!Promise}
 */
Ecdsa.verify = function(pk, hashAlg, msg, sig) {
  return crypto.subtle.verify(
      {name: 'ECDSA', hash: {name: hashAlg}}, pk, sig, msg);
};

/**
 * Signs a message using the given ECDSA private key and hash algorithm.
 * @param {!CryptoKey} sk The ECDSA private key
 * @param {!ArrayBuffer} msg The message to be signed
 * @param {string} hashAlg The hash algorithm
 *
 * @return {!Promise}
 */
Ecdsa.sign = function(sk, msg, hashAlg) {
  return crypto.subtle.sign(
      {
        name: 'ECDSA',
        hash: {name: hashAlg},
      },
      sk, msg);
};


/**
 * Imports a ECDSA public key.
 * @param {!JSONObject} keyData The key data in JWK format
 * @param {string} hashAlg The hash algorithm
 * @param {!Array<string>} usages The usages of the key
 *
 * @return {!Promise}
 */
Ecdsa.importPublicKey = function(keyData, hashAlg, usages) {
  return crypto.subtle.importKey(
      'jwk', keyData,
      {name: 'ECDSA', namedCurve: keyData['crv'], hash: {name: hashAlg}}, true,
      usages);
};

/**
 * Exports the given ECDSA key as a JSON object.
 * @param {!CryptoKey} key The ECDSA key to be exported
 *
 * @return {!Promise}
 */
Ecdsa.exportKey = function(key) {
  return crypto.subtle.exportKey('jwk', key);
};

/**
 * Generates an ECDSA key pair using the given hash algorithm and curve name.
 * @param {string} hashAlg The hash algorithm
 * @param {string} curveName The curve name
 *
 * @return {!Promise}
 */
Ecdsa.generateKey = function(hashAlg, curveName) {
  return crypto.subtle.generateKey(
      {
        name: 'ECDSA',
        namedCurve: curveName,
      },
      true, ['sign', 'verify']);
};


/**
 * Tests ECDSA signature verification. The test case's parameters are passed
 * to 'this' variable of the function.
 *
 * @return {!Promise}
 */
Ecdsa.testVerify = function() {
  tc = this;
  var promise = new Promise((resolve, reject) => {
    Ecdsa.importPublicKey(tc.keyData, tc.hashAlg, ['verify'])
        .then(function(pk) {
          Ecdsa.verify(pk, tc.hashAlg, tc.msg, tc.sig)
              .then(function(isValid) {
                if ((tc.result == 'valid' && !isValid) ||
                    (tc.result == 'invalid' && isValid)) {
                  reject('Failed in test case ' + tc.id);
                }
                resolve();
              })
              .catch(function(err) {
                // don't expect any exception in signature verification
                reject(
                    'Unexpected exception on test case ' + tc.id + ': ' + err);
              });
        })
        .catch(function(err) {
          reject('Failed to import key in test case ' + tc.id + ': ' + err);
        });
  });
  return promise;
};


/**
 * Parameters of a ECDSA signature verification test.
 * @param {!number} id Test case's id
 * @param {!JSONObject} keyData The key data in JWK format
 * @param {string} hashAlg The hash algorithm
 * @param {!ArrayBuffer} msg The message that was signed
 * @param {!ArrayBuffer} sig The signature to be verified
 * @param {string} result The expected result of the test case
 */
var EcdsaVerifyTestCase = function(id, keyData, hashAlg, msg, sig, result) {
  this.id = id;
  this.keyData = keyData;
  this.msg = msg;
  this.sig = sig;
  this.result = result;
  this.hashAlg = hashAlg;
};


/**
 * Tests ECDSA signature implementation with a number of test vectors.
 *
 * @return {!Promise}
 */
function testEcdsaVectors() {
  var tv = TestUtil.readJsonTestVectorsFromFile(ECDSA_VECTOR_FILE);
  var testCase = new goog.testing.TestCase();

  for (var i = 0; i < tv['testGroups'].length; i++) {
    var tg = tv['testGroups'][i];
    var keyData = tg['jwk'];
    var curveName = keyData['crv'];
    var hashAlg = tg['sha'];
    if (SUPPORTED['ecdsa-curve'].indexOf(curveName) == -1 ||
        SUPPORTED['hash'].indexOf(hashAlg) == -1) {
      continue;
    }
    for (var j = 0; j < tg['tests'].length; j++) {
      var tc = tg['tests'][j];
      var tcId = tc['tcId'];
      var result = tc['result'];
      var msg = TestUtil.hexToArrayBuffer(tc['msg']);
      var sig = TestUtil.hexToArrayBuffer(tc['sig']);
      var test =
          new EcdsaVerifyTestCase(tcId, keyData, hashAlg, msg, sig, result);
      testCase.addNewTest('Test ' + tcId, Ecdsa.testVerify, test);
    }
  }
  return testCase.runTestsReturningPromise().then(TestUtil.checkTestCaseResult);
}

/**
 * Extracts 'r' and 's' values from the given signature. The function assumes
 * that the given signature is the concatenation of 'r' and 's'.
 * @param {!ArrayBuffer} sig The signature
 *
 * @return {!Array<wycheproof.BigInteger>}
 */
Ecdsa.extractSig = function(sig) {
  var bytes = new Uint8Array(sig);
  var byteLen = bytes.length;
  var rBytes = bytes.subarray(0, byteLen / 2);
  var r = new BigInteger(rBytes);
  var sBytes = bytes.subarray(byteLen / 2, byteLen);
  var s = new BigInteger(sBytes);
  return [r, s];
};

/**
 * Extracts the random nonce that was used to calculate the given signature.
 * @param {!wycheproof.BigInteger} h The digest of the message
 * @param {!wycheproof.BigInteger} r The 'r' value of the signature
 * @param {!wycheproof.BigInteger} s The 's' value of the signature
 * @param {!wycheproof.BigInteger} d The private component of the ECDSA key
 * @param {!NistCurveSpec} curveSpec The specifications of the used curve
 *
 * @return {!wycheproof.BigInteger}
 */
Ecdsa.extractNonce = function(h, r, s, d, curveSpec) {
  var n = curveSpec.n;
  var k = d.multiply(r).add(h).multiply(n.modInverse(s)).mod(n);
  return k;
};

/**
 * Checks whether the given nonce was actually used during the signing process
 * of the given signature.
 * @param {!ArrayBuffer} msg The message that was signed
 * @param {!wycheproof.BigInteger} r The 'r' value of the signature
 * @param {!wycheproof.BigInteger} s The 's' value of the signature
 * @param {!wycheproof.BigInteger} d The private component of the ECDSA key
 * @param {!wycheproof.BigInteger} k The nonce that needs to be checked
 * @param {string} curveName The curve name
 */
Ecdsa.checkNonceCorrectness = function(msg, r, s, d, k, curveName) {
  var e2eCurveMap = {
    'P-256': 'P_256',
    'P-384': 'P_384',
    'P-521': 'P_521',
  };
  var e2eCurveName = e2eCurveMap[curveName];
  var key = new e2e.ecc.Ecdsa(e2eCurveName, {privKey: d.toByteArray()});
  var msgBytes = new Uint8Array(msg);
  var calSig = key.signForTestingOnly(msgBytes, k);
  var calR = new BigInteger(calSig['r']);
  var calS = new BigInteger(calSig['s']);
  assertTrue(
      'Nonce calculation was incorrect', r.isEqual(calR) && s.isEqual(calS));
};

/**
 * Tests whether there is bias in the nonce generation during
 * ECDSA signing process. The test case's parameters are passed
 * to 'this' variable of the function. The test is based on the fact that
 * if we throw a fair coin nTests times then the probability that
 * either heads or tails appears less than minCount is less than 2^{-32}.
 * Therefore the test below is not expected to fail unless the generation
 * of the one time keys is indeed biased.
 *
 * NOTE: This test only works correctly if the length of the output of the
 * hash function is equal to the bit length of the curve order.
 *
 * @return {!Promise}
 */
Ecdsa.testBias = function() {
  var tc = this;
  var countLsb = 0;
  var countMsb = 0;

  return Ecdsa.generateKey(tc.hashAlg, tc.curveName)
      .then(function(key) {
        return Ecdsa.exportKey(key.privateKey)
            .then(function(keyData) {
              var promises = [];

              for (var i = 0; i < tc.nTests; i++) {
                promises.push(
                    Ecdsa.sign(key.privateKey, tc.msg, tc.hashAlg)
                        .then(function(sig) {
                          return HashUtil.digest(tc.hashAlg, tc.msg)
                              .then(function(digest) {
                                var curveSpec =
                                    EcUtil.getCurveSpec(tc.curveName);
                                var h = new BigInteger(new Uint8Array(digest));
                                // private key value
                                var d = BigInteger.fromHex(
                                    TestUtil.base64UrlToHex(keyData['d']));
                                var r, s;
                                [r, s] = Ecdsa.extractSig(sig);
                                var k =
                                    Ecdsa.extractNonce(h, r, s, d, curveSpec);
                                // Uncomment this line to check correctness of
                                // nonce calculation
                                // Ecdsa.checkNonceCorrectness(tc.msg, r, s, d,
                                // k, tc.curveName);
                                var halfN = curveSpec.n.shiftRight(1);
                                if (k.isBitSet(0)) countLsb += 1;
                                if (k.compare(halfN) == 1) countMsb += 1;
                              });
                        }));
              }

              return Promise.all(promises).then(function() {
                if (countLsb < tc.minCount ||
                    countLsb > tc.nTests - tc.minCount) {
                  reject(
                      'Bias detected in the LSB of k' +
                      ', hash: ' + tc.hashAlg + ', curve: ' + tc.curveName +
                      ', countLSB: ' + countLsb + ', countMSB: ' + countMsb);
                }
                if (countMsb < tc.minCount ||
                    countMsb > tc.nTests - tc.minCount) {
                  reject(
                      'Bias detected in the MSB of k' +
                      ', hash: ' + tc.hashAlg + ', curve: ' + tc.curveName +
                      ', countLSB: ' + countLsb + ', countMSB: ' + countMsb);
                }
              });
            })
            .catch(function(err) {
              throw new Error('Failed to export private key: ' + err);
            });
      })
      .catch(function(err) {
        throw new Error('Failed to generate key: ' + err);
      });
};

/**
 * Parameters of a ECDSA bias test.
 * @param {string} hashAlg The hash algorithm
 * @param {string} curveName The curve name
 * @param {!ArrayBuffer} msg The message that was signed
 * @param {!number} nTests The number of key to be generated
 * @param {!number} minCount
 *   The expected number of times that a bit of the nonce should be 1 or 0
 */
var EcdsaBiasTestCase = function(hashAlg, curveName, msg, nTests, minCount) {
  this.msg = msg;
  this.hashAlg = hashAlg;
  this.curveName = curveName;
  this.nTests = nTests;
  this.minCount = minCount;
};

/**
 * Tests whether there is bias in the nonce generation during
 * ECDSA signing process. It considers three curves: secp256r1, secp384r1,
 * and secp521r1.
 *
 * @return {!Promise}
 */
function testEcdsaBiasAll() {
  var testCase = new goog.testing.TestCase();
  testCase.promiseTimeout = 120 * 1000;
  var msg = TestUtil.hexToArrayBuffer('48656c6c6f');  // msg = 'Hello'
  var nTests = 1024;
  var minCount = 410;
  var biasTest256 =
      new EcdsaBiasTestCase('SHA-256', 'P-256', msg, nTests, minCount);
  testCase.addNewTest('bias256', Ecdsa.testBias, biasTest256);
  var biasTest384 =
      new EcdsaBiasTestCase('SHA-384', 'P-384', msg, nTests, minCount);
  testCase.addNewTest('bias384', Ecdsa.testBias, biasTest384);
  var biasTest521 =
      new EcdsaBiasTestCase('SHA-512', 'P-521', msg, nTests, minCount);
  testCase.addNewTest('bias521', Ecdsa.testBias, biasTest521);
  return testCase.runTestsReturningPromise().then(
      wycheproof.TestUtil.checkTestCaseResult);
}
